scipy/optimize/minpack.py

```diff
@@ -663,10 +682,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         else:
             xdata = np.asarray(xdata)
 
-    weights = 1.0 / asarray(sigma) if sigma is not None else None
-    func = _wrap_func(f, xdata, ydata, weights)
+    # Determine type of sigma
+    if sigma is not None:
+        sigma = np.asarray(sigma)
+
+        if sigma.shape == (ydata.size, ydata.size):
+            try:
+                # scipy.linalg.cholesky requires lower=True to return L L^T = A
+                T = cholesky(sigma, lower=True)
+            except LinAlgError:
+                raise ValueError("sigma must be positive definite")
+        elif sigma.shape == (ydata.size, ):
+            T = 1.0 / sigma
+        else:
+            raise ValueError("sigma has incorrect shape")
```

Comment:
Sort of minor: I think it's good to refer to function arguments using backticks like in the docstring. That is write "\`sigma\` ..." (I finally figured how to write a backtick in markdown).